### PR TITLE
Introduce limit to the number of terms in Terms Query

### LINKED ISSUE #18829: Safeguard: limit number of terms in TermsQuery
**Describe the feature**: I have seen a lot of cases where an application will generate some long list of terms for a `TermsQuery` and pass it off to Elasticsearch. In general, `TermsQuery` performance degrades quickly as more terms are added, and once you get into hundreds (or thousands), you can pretty much forget it. (Unless there's a way to do a terms query with hundreds or thousands of terms and I just don't know the secret.) I think we should consider short-circuiting on the number of terms if it goes beyond some threshold unless some overriding argument is given. Additionally, it might be a good idea for Elastic.co documentation to warn users that many terms can have a substantial impact on latency, and of course point out the default limit/safeguard (supposing it is decided that the number of terms should be limited).

Relates #11511

### PR DESCRIPTION
**Pull request by mayya-sharipova**
- Introduce index level settings to control the maximum number of terms
    that can be used in a Terms Query
- Throw an error if a request exceeds this max number

Closes #18829

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/IndexSettings.java [hunk 1]
```diff
@@ -129,6 +129,15 @@
...
 
+
+    /**
+     * Index setting describing the maximum number of terms that can be used in Terms Query.
+     * The default maximum of 2000 terms is defensive, as extra processing and memory is involved
+     * for each additional term, and a large number of terms degrade the cluster performance.
+     */
+    public static final Setting<Integer> MAX_TERMS_COUNT_SETTING =
+        Setting.intSetting("index.max_terms_count", 2000, 1, Property.Dynamic, Property.IndexScope);
+
```
Comment (mayya-sharipova) [inline core/src/main/java/org/elasticsearch/index/IndexSettings.java]: Originally, I intended to set the limit to `1024` the same number as the Lucene maximum number of boolean clauses, but some tests in xpack use higher number of terms, thus  I had to increase this number to 2000.

**Review by jpountz — COMMENTED** [review]
The change looks good but I think we need to be careful with the default value: I suspect many users will hit it with a value of 2000. I'd be tempted to set the default value to MAX_VALUE in 6.x and maybe something higher in master, eg. 2^16. For the record, the limit on boolean queries is due to the fact that composing X term queries with a boolean query requires to stream X postings lists from disk concurrently. On the other hand, the `terms` query reads all postings lists sequentially, which is why it scales better with the number of terms (but this comes with a drawback: all matches need to be visited).

**Comment by mayya-sharipova** [comment]
@jpountz Thanks for your comments, Adrien! 
It makes sense!  I will increase limit in master to 2^16.
For the 6.x, we were also thinking to issue a deprecation warning instead of an error if the max_value is exceeded, so may be it makes sense to have the same limit as in master:  2^16?

Thanks for the explanation how boolean query  and terms query work.

**Review by jpountz — APPROVED** [review]
LGTM!

**Comment by jpountz** [comment]
> For the 6.x, we were also thinking to issue a deprecation warning instead of an error if the max_value is exceeded, so may be it makes sense to have the same limit as in master: 2^16?

Agreed.

**Comment by thomasneirynck** [comment]
Will this work transparently for clients? e.g., if Kibana issues a request for a terms-agg exceeding the threshold, will the result be automatically truncated, or will ES raise an error?

**Comment by s1monw** [comment]
@thomasneirynck it will throw an `IllegalArgumentException`

**Comment by mayya-sharipova** [comment]
@thomasneirynck for 6.2 we don't plan to throw an error unless the setting is explicitly set by a user.  So in 6.2, by default is the limit is exceeded (and the limit is 65536), only deprecation warning will be issued.  Will this create a problem in Kibana?

**Comment by thomasneirynck** [comment]
thx @mayya-sharipov. Kibana should be OK then for 6.2.

When the limit is set on the ES-side, it makes sense for admins them to make the corresponding change in Kibana.

We can expand the Kibana documentation as well.

Do you know when you will be turning off the deprecation behavior and have ES start failing?

**Comment by jpountz** [comment]
This will happen in 7.0. However I'm wondering that there might be some confusion: you mentioned `terms-agg` in your previous comment while this change is about the `terms` **query**, and I don't think Kibana should worry about it, it should just propagate the error returned by Elasticsearch to the user?

**Comment by thomasneirynck** [comment]
@jpountz you're right. my mistake, I was talking about terms-agg, not terms-query. Sorry for spam. Yeah, it will just propagate.

**Pull request merged**